plugins/linux-capture/pipewire.c

```diff
@@ -0,0 +1,1189 @@
+	bfree(call);
+}
+
+static void teardown_pipewire(obs_pipewire_data *obs_pw)
+{
+	if (obs_pw->thread_loop) {
+		pw_thread_loop_wait(obs_pw->thread_loop);
+		pw_thread_loop_stop(obs_pw->thread_loop);
+	}
+
+	if (obs_pw->stream)
+		pw_stream_disconnect(obs_pw->stream);
+	g_clear_pointer(&obs_pw->stream, pw_stream_destroy);
+	g_clear_pointer(&obs_pw->context, pw_context_destroy);
+	g_clear_pointer(&obs_pw->thread_loop, pw_thread_loop_destroy);
+
+	if (obs_pw->pipewire_fd > 0) {
+		close(obs_pw->pipewire_fd);
+		obs_pw->pipewire_fd = 0;
+	}
+
+	obs_pw->negotiated = false;
+}
+
+static void destroy_session(obs_pipewire_data *obs_pw)
+{
+	if (obs_pw->session_handle) {
+		g_dbus_connection_call(
+			obs_pw->connection, "org.freedesktop.portal.Desktop",
+			obs_pw->session_handle,
+			"org.freedesktop.portal.Session", "Close", NULL, NULL,
+			G_DBUS_CALL_FLAGS_NONE, -1, NULL, NULL, NULL);
+
+		g_clear_pointer(&obs_pw->session_handle, g_free);
+	}
+
+	g_clear_pointer(&obs_pw->sender_name, bfree);
+	g_clear_pointer(&obs_pw->cursor.texture, gs_texture_destroy);
+	g_clear_pointer(&obs_pw->texture, gs_texture_destroy);
+	g_cancellable_cancel(obs_pw->cancellable);
+	g_clear_object(&obs_pw->cancellable);
+	g_clear_object(&obs_pw->connection);
+	g_clear_object(&obs_pw->proxy);
+}
+
+static inline bool has_effective_crop(obs_pipewire_data *obs_pw)
+{
+	return obs_pw->crop.valid &&
+	       (obs_pw->crop.x != 0 || obs_pw->crop.y != 0 ||
+		obs_pw->crop.width < obs_pw->format.info.raw.size.width ||
+		obs_pw->crop.height < obs_pw->format.info.raw.size.height);
+}
+
+static bool
+spa_pixel_format_to_obs_pixel_format(uint32_t spa_format,
+				     enum gs_color_format *out_format)
+{
+	switch (spa_format) {
+	case SPA_VIDEO_FORMAT_RGBA:
+	case SPA_VIDEO_FORMAT_RGBx:
+		*out_format = GS_RGBA;
+		break;
+
+	case SPA_VIDEO_FORMAT_BGRA:
+		*out_format = GS_BGRA;
+		break;
+
+	case SPA_VIDEO_FORMAT_BGRx:
+		*out_format = GS_BGRX;
+		break;
+
+	default:
+		return false;
+	}
+
+	return true;
+}
+
+/* ------------------------------------------------- */
+
+static void on_process_cb(void *user_data)
+{
+	obs_pipewire_data *obs_pw = user_data;
+	struct spa_meta_cursor *cursor;
+	enum gs_color_format obs_format;
+	struct spa_buffer *buffer;
+	struct pw_buffer *b;
+	bool has_buffer;
+
+	/* Find the most recent buffer */
+	b = NULL;
+	while (true) {
+		struct pw_buffer *aux =
+			pw_stream_dequeue_buffer(obs_pw->stream);
+		if (!aux)
+			break;
+		if (b)
+			pw_stream_queue_buffer(obs_pw->stream, b);
+		b = aux;
+	}
+
+	if (!b) {
+		blog(LOG_DEBUG, "[pipewire] Out of buffers!");
+		return;
+	}
+
+	buffer = b->buffer;
+	has_buffer = buffer->datas[0].chunk->size != 0;
+
+	obs_enter_graphics();
+
+	if (!spa_pixel_format_to_obs_pixel_format(
+		    obs_pw->format.info.raw.format, &obs_format)) {
+		blog(LOG_ERROR, "[pipewire] unsupported buffer format: %d",
+		     obs_pw->format.info.raw.format);
+		goto read_metadata;
+	}
+
+	if (has_buffer) {
```

Comment:
Oh sorry I think it looks better similar to how you structured it before, just without the unneeded if/indention for the video crop. Also this ordering of if/else clauses makes the goto far away from the controlling if.

if (!has_buffer) { goto ... }

// unconditionally
/* Video crop */